R/backend-file.R

```diff
@@ -0,0 +1,507 @@
+  !is.null(b_file_keyring_env(private$keyring_file(keyring))$key)
+
+
+b_file_key_set <- function(self, private, key, keyring) {
+
+  key <- key %||% get_pass()
+  assert_that(is_string(key))
+  key <- sodium::hash(charToRaw(key))
+
+  kr_env <- b_file_keyring_env(private$keyring_file(keyring, password = key))
+
+  kr_env$key <- key
+}
+
+b_file_keyring_set <- function(self, private, keyring, nonce, check, items) {
+
+  kr_env <- b_file_keyring_env(private$keyring_file(keyring))
+
+  if (is.null(nonce) || is.null(check) || is.null(items))
+    kr <- private$keyring_read_file(keyring)
+
+  nonce <- nonce %||% kr[["nonce"]]
+  assert_that(is.raw(nonce), length(nonce) > 0L)
+  kr_env$nonce <- nonce
+
+  check <- check %||% kr[["check"]]
+  assert_that(is.character(check), length(check) > 0L)
+  kr_env$check <- check
+
+  kr_env$items <- lapply(items %||% kr[["items"]], b_file_validate_item)
+
+  kr_env
+}
+
+b_file_keyring_get <- function(self, private, keyring)
+  list(
+    nonce = private$nonce_get(keyring),
+    items = private$items_get(keyring),
+    check = private$check_get(keyring)
+  )
+
+b_file_nonce_get <- function(self, private, keyring) {
+
+  kr_env <- b_file_keyring_env(private$keyring_file(keyring))
+
+  if (is.null(kr_env$nonce))
+    kr_env <- private$keyring_set(keyring)
+
+  res <- kr_env$nonce
+  assert_that(is.raw(res), length(res) > 0L)
+
+  res
+}
+
+b_file_items_get <- function(self, private, keyring) {
+
+  kr_env <- b_file_keyring_env(private$keyring_file(keyring))
+
+  if (is.null(kr_env$items))
+    private$keyring_set(keyring)
+
+  lapply(kr_env$items, b_file_validate_item)
+}
+
+b_file_check_get <- function(self, private, keyring) {
+
+  kr_env <- b_file_keyring_env(private$keyring_file(keyring))
+
+  if (is.null(kr_env$check))
+    private$keyring_set(keyring)
+
+  res <- kr_env$check
+  assert_that(is.character(res), length(res) > 0L)
+
+  res
+}
+
+## --------------------------------------------------------------------
+## helper functions
+
+b_file_secret_encrypt <- function(secret, nonce, key) {
+
+  res <- sodium::data_encrypt(
+    charToRaw(secret),
+    key,
+    nonce
+  )
+
+  b_file_split_string(sodium::bin2hex(res))
+}
+
+b_file_secret_decrypt <- function(secret, nonce, key)
+  rawToChar(
+    sodium::data_decrypt(
+      sodium::hex2bin(b_file_merge_string(secret)),
+      key,
+      nonce
+    )
+  )
+
+b_file_keyring_env <- function(keyring) {
+
+  env_name <- normalizePath(keyring, mustWork = TRUE)
+
+  kr_env <- get0(env_name, envir = keyrings, mode = "environment")
+
+  if (is.null(kr_env))
+    kr_env <- assign(env_name, new.env(), envir = keyrings)
+
+  kr_env
+}
+
+b_file_error <- function(problem, reason = NULL) {
+  info <- if (is.null(reason))
+    problem
+  else
+    paste0(problem, ": ", reason)
+  stop("keyring error (file-based keyring), ", info, call. = FALSE)
+}
+
+b_file_validate_item <- function(item) {
+
+  assert_that(is.list(item), length(item) == 3L,
```

Comment:
Again, a better error message would be great.

Reply:
done (6bb91cd4fd921d040035c73c170bb24c2232cab0 and improved 17c615702c5a85cfc65c9bf9c259a7397ff06513)